iquip/apps/dataviewer.py

```diff
@@ -634,24 +733,213 @@ def datasetName(self) -> str:
         """Extended."""
         super().__init__(name, parent=parent)
         self.frame = DataViewerFrame()
+        self.thread: Optional[_DatasetFetcherThread] = None
         self.policy: Optional[SimpleScanDataPolicy] = None
+        self.axis: Tuple[int, ...] = ()
+        self.dataPointIndex: Tuple[int, ...] = ()
+        self.frame.syncRequested.connect(self.synchronize)
+        self.frame.sourceWidget.axisApplied.connect(self.setAxis)
+        self.frame.dataPointWidget.dataTypeChanged.connect(self.setDataType)
+        self.frame.dataPointWidget.thresholdChanged.connect(self.setThreshold)
+        self.frame.mainPlotWidget.dataClicked.connect(self.selectDataPoint)
+
+    @pyqtSlot()
+    def synchronize(self):
+        """Fetches the dataset from artiq master and updates the viewer."""
+        self.thread = _DatasetFetcherThread(
+            self.frame.datasetName(),
+            self.constants.proxy_ip,  # pylint: disable=no-member
+            self.constants.proxy_port,  # pylint: disable=no-member
+            self.setDataset,
+        )
+        self.thread.start()
+
+    @pyqtSlot(np.ndarray, list, list)
+    def setDataset(
+        self,
+        dataset: np.ndarray,
+        parameters: List[str],
+        units: List[Optional[str]],
+    ):
+        """Sets the dataset to show and updates the axis viewer.
+        
+        Args:
+            See SimpleScanDataPolicy.
+        """
+        self.policy = SimpleScanDataPolicy(dataset, parameters, units)
+        self.frame.sourceWidget.setParameters(parameters, units)
+
+    @pyqtSlot(tuple)
+    def setAxis(self, axis: Sequence[int]):
+        """Given the axis information, draws the main plot.
+        
+        Args:
+            axis: See updateMainPlot().
+        """
+        self.axis = axis
+        dataType = self.frame.dataPointWidget.dataType()
+        self.updateMainPlot(axis, dataType)
+
+    @pyqtSlot(DataPointWidget.DataType)
+    def setDataType(self, dataType: DataPointWidget.DataType):
+        """Given the data type, draws the main plot.
+        
+        Args:
+            dataType: See updateMainPlot().
+        """
+        self.updateMainPlot(self.axis, dataType)
+
+    def updateMainPlot(self, axis: Sequence[int], dataType: DataPointWidget.DataType):
+        """Updates the main plot.
+        
+        Args:
+            axis: See SimpleScanDataPolicy.extract().
+            dataType: Target data type.
+        """
+        if self.policy is None:
+            return
+        reduce = self._reduceFunction(dataType)
+        data, axes = self.policy.extract(axis, reduce)
+        self.frame.mainPlotWidget.setData(data, axes)
+        index = self.dataPointIndex
+        if data.ndim == len(index) and np.all(np.less(index, data.shape)):
+            self.selectDataPoint(index)
+        else:
+            self.selectDataPoint((0,) * data.ndim)
+
+    @pyqtSlot(tuple)
+    def selectDataPoint(self, index: Tuple[int, ...]):
```

Comment:
Why didn't you update the number of samples here?

I think you can update it as follows:
```python
self.frame.dataPointWidget.setNumberOfSamples(len(data))
```

Reply:
You are right! I totally forgot about that.